src/v/cloud_storage/remote_partition.cc

```diff
@@ -670,8 +676,17 @@ void remote_partition::return_reader(
       [this, &reader](offloaded_segment_state&) {
           evict_reader(std::move(reader));
       },
-      [&reader](materialized_segment_ptr& m_state) {
-          m_state->return_reader(std::move(reader));
+      [this, &reader](materialized_segment_ptr& m_state) {
+          if (_readers_limit.current() <= 0) {
```

Comment:
Possibly should be changed to `available_units()` if we want to check for the semaphore going negative.